File: answers/24/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>
#include "../hist.cpp"

void gamma_corr(const cv::Mat &src, cv::Mat &dst, float c, float g) {
    dst = src.clone();
    dst.convertTo(dst, CV_32F);

    dst.forEach<cv::Vec3f>([&](cv::Vec3f &pixel, const int pos[]) -> void {
        for (int i=0; i<3; i++) {
            pixel[i] = pow(((1 / c) * (pixel[i] / 255.)), (1 / g)) * 255;
        }
    });
    dst.convertTo(dst, CV_8U);
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256_gamma.png");
    cv::Mat img_result;
    gamma_corr(img_orig, img_result, 1, 2.2);

    cv::Mat hist_orig, hist_result;
    makeHisto(img_orig, hist_orig, 256, 0, 256);
    makeHisto(img_result, hist_result, 256, 0, 256);
    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::imshow("cpp_origin_hist", hist_orig);
    cv::imshow("cpp_result_hist", hist_result);
    cv::moveWindow("cpp_origin", 0, img_orig.size().height);
    cv::moveWindow("cpp_result", img_orig.size().width, img_orig.size().height);
    cv::moveWindow("cpp_origin_hist", img_orig.size().width + img_result.size().width, img_orig.size().height);
    cv::moveWindow("cpp_result_hist", img_orig.size().width + img_result.size().width + hist_orig.size().width, img_orig.size().height);
    cv::waitKey(0);
    return 0;
}

File: answers/24/main.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
import sys
from answers.hist import *

def gamma_corr(img, c, g):
    out = (1 / c * (img / 255)) ** (1 / g)
    return np.clip(out * 255, 0, 255).astype(np.uint8)

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256_gamma.png')
    img_result = gamma_corr(img_orig, c=1, g=2.2)

    hist_orig = makeHisto(img_orig, 256, 0, 256)
    hist_result = makeHisto(img_result, 256, 0, 256)

    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.imshow('python_origin_hist', hist_orig)
    cv2.imshow('python_result_hist', hist_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.moveWindow('python_origin_hist', img_orig.shape[1]+img_result.shape[1], 0)
    cv2.moveWindow('python_result_hist', img_orig.shape[1]+img_result.shape[1]+hist_orig.shape[1], 0)
    cv2.waitKey(0)
